crates/scheduler/src/precedence.rs

```diff
@@ -0,0 +1,276 @@
+        let read_system1 = into_system(other_read_a);
+
+        let ordering0 = read_system0.precedence_to(read_system1.as_ref());
+        let ordering1 = read_system1.precedence_to(read_system0.as_ref());
+
+        assert_eq!(Precedence::Equal, ordering0);
+        assert_eq!(Precedence::Equal, ordering1);
+    }
+
+    #[test]
+    fn system_reading_from_many_and_writing_to_component_precedes_system_read_from_component() {
+        let many_system = into_system(read_b_write_a);
+        let read_system = into_system(read_a);
+
+        let ordering0 = many_system.precedence_to(read_system.as_ref());
+        let ordering1 = read_system.precedence_to(many_system.as_ref());
+
+        assert_eq!(Precedence::Before, ordering0);
+        assert_eq!(Precedence::After, ordering1);
+    }
+
+    #[test]
+    fn system_reading_from_many_and_writing_to_component_precedes_system_writing_to_many_and_reading_from_component(
+    ) {
+        let many_parameters_but_single_write = into_system(read_b_write_a);
+        let many_parameters_but_single_read = into_system(read_a_write_c);
+
+        let ordering0 = many_parameters_but_single_write
+            .precedence_to(many_parameters_but_single_read.as_ref());
+        let ordering1 = many_parameters_but_single_read
+            .precedence_to(many_parameters_but_single_write.as_ref());
+
+        assert_eq!(Precedence::Before, ordering0);
+        assert_eq!(Precedence::After, ordering1);
+    }
+
+    #[test]
+    fn system_writing_precedes_system_reading_even_if_both_also_read_other_component() {
+        let read_a_write_b = into_system(read_a_write_b);
+        let read_ab = into_system(read_ab);
+
+        let ordering0 = read_a_write_b.precedence_to(read_ab.as_ref());
+        let ordering1 = read_ab.precedence_to(read_a_write_b.as_ref());
+
+        assert_eq!(Precedence::Before, ordering0);
+        assert_eq!(Precedence::After, ordering1);
+    }
+
+    fn arb_systems() -> Vec<Box<dyn System>> {
+        vec![
+            into_system(read_a),
+            into_system(other_read_a),
+            into_system(write_a),
+            into_system(other_write_a),
+            into_system(read_a_write_b),
+            into_system(read_ab),
+        ]
+    }
+
+    prop_compose! {
+        fn arb_system(system_count: usize)
+                     (system_index in 0..system_count)
+                     -> Box<dyn System> {
+            let mut systems = arb_systems();
+            systems.remove(system_index)
+        }
+    }
+
+    prop_compose! {
+        fn arb_ordered_systems(max_system_count: usize)
+                              (systems_count in 1..=max_system_count)
+                              (systems in hash_set(arb_system(systems_count), systems_count))
+                              -> HashSet<Box<dyn System>> {
+            systems
+        }
+    }
+
+    /// NOTE: this assertion does _not_ hold for systems that mutually read and write to each others
```

Comment:
Good note! I was thinking about this scenario when reading the implementation and this explains it.